src/world/Units/Unit.cpp

```diff
@@ -76,167 +77,409 @@ void Unit::setClass(uint8_t class_) { write(unitData()->field_bytes_0.s.unit_cla
 uint8_t Unit::getGender() const { return unitData()->field_bytes_0.s.gender; }
 void Unit::setGender(uint8_t gender) { write(unitData()->field_bytes_0.s.gender, gender); }
 
-uint8_t Unit::getPowerType() const { return unitData()->field_bytes_0.s.power_type; }
-void Unit::setPowerType(uint8_t powerType) { write(unitData()->field_bytes_0.s.power_type, powerType); }
+PowerType Unit::getPowerType() const { return static_cast<PowerType>(unitData()->field_bytes_0.s.power_type); }
+void Unit::setPowerType(uint8_t powerType)
+{
+    write(unitData()->field_bytes_0.s.power_type, powerType);
+
+#ifdef AE_TBC
+    // TODO Fix this later
+    return;
+#endif
+
+    // Update power type also to group
+    const auto plr = getPlayerOwner();
+    if (plr == nullptr || !plr->IsInWorld() || plr->GetGroup() == nullptr)
+        return;
+
+    plr->AddGroupUpdateFlag(isPlayer() ? GROUP_UPDATE_FLAG_POWER_TYPE : GROUP_UPDATE_FLAG_PET_POWER_TYPE);
+}
 //bytes_0 end
 
 uint32_t Unit::getHealth() const { return unitData()->health; }
-void Unit::setHealth(uint32_t health) { write(unitData()->health, health); }
+void Unit::setHealth(uint32_t health)
+{
+    const auto maxHealth = getMaxHealth();
+    if (health > maxHealth)
+        health = maxHealth;
+
+    write(unitData()->health, health);
+
+#ifdef AE_TBC
+    // TODO Fix this later
+    return;
+#endif
+
+    // Update health also to group
+    const auto plr = getPlayerOwner();
+    if (plr == nullptr || !plr->IsInWorld() || plr->GetGroup() == nullptr)
+        return;
+
+    plr->AddGroupUpdateFlag(isPlayer() ? GROUP_UPDATE_FLAG_CUR_HP : GROUP_UPDATE_FLAG_PET_CUR_HP);
+}
 void Unit::modHealth(int32_t health)
 {
-    uint32_t currentHealth = getHealth();
-    currentHealth += health;
-    setHealth(currentHealth);
+    int32_t newHealth = getHealth();
+    newHealth += health;
+
+    if (newHealth < 0)
+        newHealth = 0;
+
+    setHealth(newHealth);
 }
 
-uint32_t Unit::getPower(uint16_t index) const
+uint32_t Unit::getPower(PowerType type) const
 {
-    switch (index)
+    if (type == POWER_TYPE_HEALTH)
+        return getHealth();
+
+    // Since cata power fields work differently
+    // Get matching power index by power type
+    const auto powerIndex = getPowerIndexFromDBC(type);
+    switch (powerIndex)
     {
-        case POWER_TYPE_MANA:
+        case 1:
```

Comment:
Attention, "magic numbers" are not the best solution if we could have some enum values. If there is no fitting enum for this usecase, we should create one.

Reply:
I've created a new enum to UnitDefines